fastocr.py

```diff
@@ -32,49 +32,62 @@ def setupColour(prefix, outputDict, digitList):
         if IMAGE_MULT != 1:
             img = img.resize((IMAGE_SIZE*IMAGE_MULT,
                               IMAGE_SIZE*IMAGE_MULT),PIL.Image.ANTIALIAS)
-        img = img.getdata()
         if NP_SUPPORTED:
+            img = img.getdata()
             img = np.asarray(img)
             img = np.reshape(img, (IMAGE_SIZE * IMAGE_MULT, IMAGE_SIZE*IMAGE_MULT))
+        else:
+            img = img.load()
+
         outputDict[digit] = img
         
 def setupData():
     setupColour('sprite_templates/', data, digitsLetters) #setup white
     setupColour('sprite_templates/red', redData, digits) #setup red
 
-def dist(col):
-    return col*col
-    
-def sub(col1,col2):
-    return col1-col2
-
 
-def scoreDigit(img, pattern, startX, startY, red):
-    
-    scores = []
+def getDigit(img, pattern, startX, startY, red):
     template = redData if red else data
-    validDigits = digitsLetters if pattern == 'A' else digits    
+    validDigits = digitsLetters if pattern == 'A' else digits
+
     if NP_SUPPORTED:
+        scores = {}
         #img in y, x format
-        subImage = img[:,startX:startX + 14]    
-    for digit in validDigits:
-        score = 0
-        if NP_SUPPORTED:                        
-            diff = np.subtract(subImage, template[digit])            
+        subImage = img[:,startX:startX + 14]
+
+        for digit in validDigits:
+            diff = np.subtract(subImage, template[digit])
             diff = np.abs(diff)
-            score = np.sum(diff)  
-            
-        else:
-            for y in range(IMAGE_SIZE*IMAGE_MULT):
-                for x in range(IMAGE_SIZE*IMAGE_MULT):
-                    a = template[digit][x,y]
-                    b = img[startX+x,startY+y]
-                    score += dist(sub(a,b))
-                
-        scores.append((score, digit))
-    scores.sort(key=lambda tup:tup[0])
-    
-    return scores[0]
+            scores[digit] = np.sum(diff)
+
+    else:
+        scores = {digit:0 for digit in validDigits}
+        MAX = IMAGE_SIZE * IMAGE_MULT
+
+        y = 0
+        while y < MAX:
+            x = 0
+            while x < MAX:
```

Comment:
being "pythonic" here, its better to use `y in range(MAX)` and `x in range(MAX)` rather than `while -> x+=1`. Is it actually faster to use manual loop iteration?

Reply:
Thanks for making me verify that. The pythonic way looks better is faster too, nearly twice!

Simple benchmark here:
https://gist.github.com/timotheeg/e06f042f09b1d50c5bdac4fe00c77da5

I've updated the commit to restore the pythonic style 😃